src/hotspot/share/runtime/atomic.hpp

```diff
@@ -786,7 +786,7 @@ inline T Atomic::cmpxchg_using_helper(Fn fn,
   return PrimitiveConversions::cast<T>(
     fn(PrimitiveConversions::cast<Type>(exchange_value),
        reinterpret_cast<Type volatile*>(dest),
-       PrimitiveConversions::cast<Type>(compare_value)));
+       PrimitiveConversions::cast<Type>(compare_value),fn));
 }
```

Comment:
After reviewing this change carefully, I don't believe it is the correct change. I reverted it and took a look at the compilation errrors on OpenBSD/i386 with clang 7.0.1. These are the relevant error messages:
```
/home/truk/jdk/openjdk-jdk11u/src/hotspot/share/runtime/atomic.hpp:789:55: error: too few arguments to function call, expected 4, have 3
       PrimitiveConversions::cast<Type>(compare_value)));
                                                      ^
/home/truk/jdk/openjdk-jdk11u/src/hotspot/os_cpu/bsd_x86/atomic_bsd_x86.hpp:150:10: note: in instantiation of function template specialization 'Atomic::cmpxchg_using_helper<long long, long long (*)(long long, volatile long long *, long long, bool), long long>' requested here
  return cmpxchg_using_helper<int64_t>(_Atomic_cmpxchg_long, exchange_value, dest, compare_value);
         ^
/home/truk/jdk/openjdk-jdk11u/src/hotspot/share/runtime/atomic.hpp:721:12: note: in instantiation of function template specialization 'Atomic::PlatformCmpxchg<8>::operator()<long long>' requested here
    return PlatformCmpxchg<sizeof(T)>()(exchange_value,
           ^
/home/truk/jdk/openjdk-jdk11u/src/hotspot/share/runtime/atomic.hpp:697:10: note: in instantiation of member function 'Atomic::CmpxchgImpl<long long, long long, long long, void>::operator()' requested here
  return CmpxchgImpl<T, D, U>()(exchange_value, dest, compare_value, order);
         ^
/home/truk/jdk/openjdk-jdk11u/src/hotspot/share/oops/accessBackend.inline.hpp:196:18: note: in instantiation of function template specialization 'Atomic::cmpxchg<long long, long long, long long>' requested here
  return Atomic::cmpxchg(new_value,
                 ^
```
Everything looks good to me until atomic_bsd_x86.hpp:150, which looks like this:
```
template<>
template<typename T>
inline T Atomic::PlatformCmpxchg<8>::operator()(T exchange_value,
                                                T volatile* dest,
                                                T compare_value,
                                                atomic_memory_order /* order */) const {
  STATIC_ASSERT(8 == sizeof(T));
  return cmpxchg_using_helper<int64_t>(_Atomic_cmpxchg_long, exchange_value, dest, compare_value);
}
```
cmpxchg_using_helper will call _Atomic_cmpxchg_long with 3 arguments, however _Atomic_cmpxchg_long is defined with 4 arguments on line 139:
```
#else // !AMD64

extern "C" {
  // defined in bsd_x86.s
  int64_t _Atomic_cmpxchg_long(int64_t, volatile int64_t*, int64_t, bool);
  void _Atomic_move_long(const volatile int64_t* src, volatile int64_t* dst);
}
```
This looks like the bsd x86_32 code has fell behind and is suffering a bit of bit rot, because the fourth parameter was removed for Linux in this jdk10 change, but the bsd one was missed:

https://hg.openjdk.java.net/jdk10/jdk10/hotspot/rev/ae91ec8b554a#l2.128

It looks like Apple is no longer building the jdk for 32bit so we will need to be on the lookout for this kind of bit rot in the bsd i386 code base. I'll submit a pull request that reverts this change and implements the removal of the `bool is_MP` forth parameter which I believe is correct and matches the Linux version of this method.

Reply:
Thank you, Kurt, for the effort to clarify the context so accurately! I've decided to listen to my new compiler friend first:
# ln -s ...
# Kurt --version
.. Version 7.0.1 (Tags / RELEASE_701 / final)
target: i386-portbld-freebsd12.0 ;-)